packages/messaging/android/src/main/java/io/invertase/firebase/messaging/ReactNativeFirebaseMessagingStoreImpl.java

```diff
@@ -46,7 +56,39 @@ public RemoteMessage getFirebaseMessage(String remoteMessageId) {
 
   @Override
   public void clearFirebaseMessage(String remoteMessageId) {
-    UniversalFirebasePreferences.getSharedInstance().remove(remoteMessageId);
+    UniversalFirebasePreferences preferences = UniversalFirebasePreferences.getSharedInstance();
+    preferences.remove(remoteMessageId);
+    // check and remove old notifications message
+    String notifications = preferences.getStringValue(S_KEY_ALL_NOTIFICATION_IDS, "");
+    if (!notifications.isEmpty()) {
+      notifications = notifications.replace(remoteMessageId + ",", ""); // remove from list
+    }
+    List<String> allNotificationList = convertToArray(notifications);
+    if (allNotificationList.size() > MAX_SIZE_NOTIFICATIONS) {
```

Comment:
Ah - so a "ring buffer", limited to 100 now (seems like a lot! but it is bounded at least, that is the most important part) - the second part of the logic though, where you check if there are too many and we need to remove one, should be done on storeFirebaseMessage above shouldn't it?

This approach seems like it can work though...

Reply:
@mikehardy as you known, we can receive notification at many times, but we don't known what time notification item will be removed from status bar. then i will store all messages id we received from fcm and check and removed old notification when user trigger tap any item from status bar.

Reply:
But this is the exact overflow case... It is brutal, but we must not overflow, so the limit needs to apply on message ingress I think

100 visible notifications should be more than sufficient for a well behaved app? And the API should just carry a note that you will lose notification attribution if you have more than that, so the app handles it.

If you just remove on delete, it's possible for an ill-behaved app to overflow in the current style